### DISCUSSION TITLE
toga webview load local html file

### DISCUSSION THREAD

**Discussion by Jzhenli** [beeware/toga#2750] [Q&A] [answered]
Hi, if it's possible for webview widget to load local html file?  In some case we want to display a customized page before loading the remote web url.

**Comment by mhsmith** [comment, marked as answer]
As it says in [the documentation](https://toga.readthedocs.io/en/stable/reference/api/widgets/webview.html):

> Due to app security restrictions, WebView can only display http:// and https:// URLs, not file:// URLs. To serve local file content, run a web server on localhost using a background thread.

This could be as simple as using the standard library `http.server` module as shown [here](https://stackoverflow.com/a/69088143).

**Reply by Jzhenli** [reply]
@mhsmith thanks for your reply. Another little question, what's the web engine for toga webview widget on android, seems it's not mentioned in the doc.

**Reply by mhsmith** [reply]
It's the standard Android WebView.

**Reply by freakboy3742** [reply]
For completeness - another option for loading local content would be to read the HTML content as a text file, and set the HTML directly on the webview with `set_content()`. 

However, this only works if the HTML content is *entirely* self contained, or only contains references to external URLs. Local URL references won't resolve, so if your web page contains a reference to a separate CSS page, or to images, that won't work - you'll need to use the local web server approach.

**Reply by Jzhenli** [reply]
> For completeness - another option for loading local content would be to read the HTML content as a text file, and set the HTML directly on the webview with `set_content()`.
> 
> However, this only works if the HTML content is _entirely_ self contained, or only contains references to external URLs. Local URL references won't resolve, so if your web page contains a reference to a separate CSS page, or to images, that won't work - you'll need to use the local web server approach.

@freakboy3742 thanks for your reply, I use the set_content() to load html, but I don't know how to change webview widget to remote url automatically after some seconds when the main_window.show() have run, if there are some call back method to do this? In my case, the remote url maybe very slow, so I want to display a customized html before loading.

**Reply by mhsmith** [reply]
In that case you'd probably need two WebViews, a visible one showing the local HTML, and an invisible one loading the remote URL, and then switch them when the remote URL has loaded.

For a fixed delay, you could use [`App.loop`](https://toga.readthedocs.io/en/stable/reference/api/app.html#toga.App.loop) to call [`call_later`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_later) for a regular function, or [`create_task`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_task) for an async function which starts off with an async sleep.

Even better would be to use use the [`on_webview_load`](https://toga.readthedocs.io/en/stable/reference/api/widgets/webview.html#toga.WebView.on_webview_load) event, but unfortunately that  isn't supported on Android yet.

**Reply by Jzhenli** [reply]
@mhsmith Hi, I try to load a local html file in the android webview, but it can not display, not sure if it's a bug?  attached is my test  html file.
[html.zip](https://github.com/user-attachments/files/18840781/html.zip)

**Reply by freakboy3742** [reply]
Based on the detail you've provided, it's impossible to tell if it is "not displaying" because of how you're using the widget. 

However, assuming you're using `WebView.set_content()` correctly, there's very little we can do to debug what is going on. We don't have any insight into the internals of the WebView - it's a widget provided by Android, and we're using it. If you can display a simple `<h1>Hello World</h1>` page, but your HTML content doesn't load, then you're at the point of diagnosing either your own HTML, or the limitations of Android's webview.

**Reply by Jzhenli** [reply]
@freakboy3742 I have find the root cause, I have changed the "**color: #f00**" to "**color: %23f00**" in the html file, then it will display well.  So I doubt in the WebView.set_content(), maybe toga have implemented it with loaddata(), from the [android api](https://www.android-doc.com/reference/android/webkit/WebView.html#loadData(java.lang.String,%20java.lang.String,%20java.lang.String)), we can find that when html file have  '#', '%', '\', '?' should be replaced by %23, %25, %27, %3f respectively. If you implemented it with loadDataWithBaseURL() maybe avoid this issue. 
<img width="672" alt="image" src="https://github.com/user-attachments/assets/b3da071c-4e19-448c-b62c-d2f098301993" />

**Reply by freakboy3742** [reply]
Ah - in which case, that's a known issue: see #2242. The proposed fix on that ticket is to use Python to encode the payload passed to `loadData()` - but `loadDataWithBaseURL()` may also be a viable fix. PRs to address the problem are welcome.
